### DISCUSSION TITLE
Feature Request: A change history embedded in each dist index.html page (Sample code in link)

### DISCUSSION THREAD

**Discussion by ModelEarth** [observablehq/framework#1040] [Ideas]
By default, when `yarn dev` or `npm run dev` are run, only update the date and time in the index.html "dist" files when an index.html file has changed. Changing the Update date in the footer when the content has not changed causes a lot of unnecessary file storage in GitHub. Also provide the option to store a Change History in each index.html page, and display when clicking a "Changed" link that reside before the most recent change date. Also provide the option to display a "Built" date - and what the heck, a Built History option while we're at it.

ChatGPT code sample removed per Fil's request. Thanks for the footer info Fil!

New code for Change History is described and linked to at the bottom.
(Currently just Change History, needs revision to include Built History too.)
Also here's the output (see anything to tweak?):
https://chat.openai.com/share/cef80ef6-7248-4a30-a93e-780482bd6e50

**Comment by Fil** [comment]
I believe your request is about the build script (as the dev script does not modify files in dist/). Even if nothing changes it might be important for a reader to know that the site has been built recently—that they are not looking at a stale page. If you don't want the page to show the date in the footer, you could specify a different **footer** option (see [config#footer](https://observablehq.com/framework/config#footer)).

(Also please don't share code generated by ChatGPT as it just muddies the discussion.)

**Comment by mbostock** [comment]
To reiterate @Fil’s answer, you just need to set the [**footer** config option](https://observablehq.com/framework/config#footer) to not show the current date in the footer. For example:

```js
export default {
  footer: `Built with <a href="https://observablehq.com/" target="_blank">Observable</a>.`
};
```

If you _want_ the date in the footer, but you only want to update it when the data or the code has changed, that’s a harder problem. I guess you’d need to look at the modification times of the Markdown page, or its transitive sources, or perhaps you’d need to keep the last build around in a cache and check if the contents of the data have changed. I’m sure someone could figure out if they really wanted, but I think as a default it’s best to show the most recent time the page was built, even if the contents haven’t changed.

**Reply by ModelEarth** [reply]
Would be great to show two optional times in the footer: Built and Changed.
And really cool to store a change history in an array of objects stored in the page (up to a 2500 characters).
With a link on "Changed" to reveal the Change History in a table below the footer links.

Here are documented steps, let me know if you see any needed modifications.

In Observable Framework, when "yarn dev" is run, if the observablehq.config.ts footer line contains an A HREF tag with id="changed", store a temporary cache of the index.html pages pulled from the project's "dist" folder. 

Example of line in observablehq.config.ts:
footer: 'Built with <a href="https://observablehq.com/" target="_blank">Observable</a> and <a id="changed" title="2024-03-11T16:36:05">Changed</a> Mar 11, 2024.'

Store the path to the files for each cached index.html page for later comparison to the new index.html files.

After the new files in "dist" are generated by "yarn dev", compare the older index.html content in the cache with the new index.html pages in the "dist" folder. Ignore the content of the footer tag when generating a string that conveys the modifications found in the new index.html version.

Parse the existing change history from the older cached index.html pages and retain the change dates when storing the prior change histories.

Prepend the most recent change to the prior change history. Include the new change date and the prior dates for each change.

Truncate the oldest when the entire change history exceeds 2500 characters. 

Insert the updated change history into the index.html page footer tag after the footer links in the "dist" folder.Some pages will have no changes. Do not update pages that had no changes.

If it does not already exist in the index.html being updated, insert javascript that reveals the history of changes below the footer links when the user clicks "Changed". Display the most recent changes at the top of the stack. Provide an "Oldest first" link above the list to sort in the reverse order with oldest first. When "Oldest" is clicked, provide a "Newest first" link to sort back the initial order.

When the user clicks "Changed" again, hide the change history. Reveal again if the Change History is currently hidden.

Here's the resulting code to add a Change History when clicking "Changed" by the change date string in the footer.
Cheerio is recommended:
https://chat.openai.com/share/cef80ef6-7248-4a30-a93e-780482bd6e50
